shared/subspace-data-retrieval/src/segment_fetcher.rs

```diff
@@ -85,66 +80,87 @@ where
 {
     debug!(%segment_index, "Retrieving pieces of the segment");
 
-    let semaphore = &Semaphore::new(RecordedHistorySegment::NUM_RAW_RECORDS);
-
-    let mut received_segment_pieces = segment_index
-        .segment_piece_indexes_source_first()
-        .into_iter()
-        .map(|piece_index| {
-            // Source pieces will acquire permit here right away
-            let maybe_permit = semaphore.try_acquire();
-
-            async move {
-                let permit = match maybe_permit {
-                    Some(permit) => permit,
-                    None => {
-                        // Other pieces will acquire permit here instead
-                        semaphore.acquire().await
-                    }
-                };
-                let piece = match piece_getter.get_piece(piece_index).await {
-                    Ok(Some(piece)) => piece,
-                    Ok(None) => {
-                        trace!(?piece_index, "Piece not found");
-                        return None;
-                    }
-                    Err(error) => {
-                        trace!(
-                            %error,
-                            ?piece_index,
-                            "Piece request failed",
-                        );
-                        return None;
-                    }
-                };
-
-                trace!(?piece_index, "Piece request succeeded");
-
-                // Piece was received successfully, "remove" this slot from semaphore
-                permit.forget();
-                Some((piece_index, piece))
-            }
-        })
-        .collect::<FuturesUnordered<_>>();
-
+    // We want NUM_RAW_RECORDS pieces to reconstruct the segment, but it doesn't matter exactly which ones.
+    let piece_indexes = segment_index.segment_piece_indexes_source_first();
+    let mut piece_indexes = piece_indexes.as_slice();
     let mut segment_pieces = vec![None::<Piece>; ArchivedHistorySegment::NUM_PIECES];
+
+    let mut pieces_pending = 0;
     let mut pieces_received = 0;
+    let mut piece_streams = Vec::new();
```

Comment:
Why do you need a vector? I don't think I understand what is going on.

Reply:
The vector stores a list of piece fetching batches. The batches get started and polled concurrently for efficiency.